Describe setting to infer labels. (#1620)
neo4j/docs-operations: modules/ROOT/pages/configuration/configuration-settings.adoc

|===

[[config_dbms.cypher.infer_schema_parts]]
=== `dbms.cypher.infer_schema_parts`

> .dbms.cypher.infer_schema_parts

Reviewer: The same should be added to the code base.

Author: This is what I try to get in:
https://github.com/neo-technology/neo4j/pull/25555/files#diff-22e5a6968aee5ed485e16da8650e784a16dbd787005216bb719d3dc44aa2b43aR249

Author: That PR has been merged but the check still fails.

[frame="topbot", stripes=odd, grid="cols", cols="<1s,<4"]
|===
|Description
a|For some queries, the planner can infer predicates such as labels or types from the graph structure that can improve estimating the number of rows that each operator produces.
(See link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/planning-and-tuning/execution-plans/#runtimes-reading-execution-plans[Cypher Manual -> Execution plans and query tuning -> Understanding execution plans])
This setting controls to what extent it should do that:

* `OFF`: No predicates are inferred.

> * `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes and the label corresponding to the least number of nodes is used to estimate rows.

Reviewer: ```suggestion
* `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes, and the label corresponding to the least number of nodes is used to estimate rows.
```

Author: ```suggestion
* `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes.
The label corresponding to the least number of nodes is used to estimate rows.
```

> Not inferring more labels yields better results for nodes with several labels which are dependent, such as every `:Actor` being a `:Person`.

Reviewer: ```suggestion
Not inferring more labels yields better results for nodes with several dependent labels, such as every `:Actor` being a `:Person`.
```

> // In general, inferring more information should improve the estimation and thereby the planner's decisions.

Reviewer: The comments should also be added to the codebase if we want them here; otherwise, the test will fail.

Author: I am not sure I follow what test would fail.

The idea behind the text in the comment was that it only makes sense to me once [we have flipped the default from `OFF` to `MOST_SELECTIVE_LABEL`.](https://trello.com/c/2IqnQCAc/2275-turn-on-label-inference-by-default)

My idea was that I could already phrase things so that the next developer could just enable the text. But I did not want to commit just yet, as toggling the feature flag might slip the release.

Reviewer: We are testing all settings to make sure that all documented valid values and default values are up-to-date with the codebase. If there are mismatches, then the test fails. I am currently working on fixing the mismatches in the descriptions so we can test them as well. In other words, the documentation should reflect the code literary.

For example, the failing test below shows that there is no such configuration setting in the codebase.

|===

[[config_dbms.cypher.infer_schema_parts]]
=== `dbms.cypher.infer_schema_parts`

.dbms.cypher.infer_schema_parts
[frame="topbot", stripes=odd, grid="cols", cols="<1s,<4"]
|===
|Description

> a|Allow label inference during cardinality estimation. During cardinality estimation, if the planner can logically deduce that a node has a label that was not explicitly expressed in the query, the planner will use this information during cardinality estimation. +

Reviewer: ```suggestion
a|Allow label inference during cardinality estimation. If the planner can logically deduce that a node has a label not explicitly expressed in the query, the planner will use this information during cardinality estimation.
```

This setting controls to what extent the planner should do that:

* `OFF`: No predicates are inferred.

> * `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes. The planner only infers at most one label per node. If more than one label can be inferred for a given node, the planner keeps the most selective one, the one corresponding to the smallest number of nodes in the graph.

Reviewer: ```suggestion
* `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes.
The planner only infers at most one label per node.
If more than one label can be inferred for a given node, the planner keeps the most selective one, the one corresponding to the smallest number of nodes in the graph.
```

|===

[[config_dbms.cypher.infer_schema_parts]]
=== `dbms.cypher.infer_schema_parts`

.dbms.cypher.infer_schema_parts
[frame="topbot", stripes=odd, grid="cols", cols="<1s,<4"]
|===
|Description

> a|Allow label inference during cardinality estimation. During cardinality estimation, if the planner can logically deduce that a node has a label that was not explicitly expressed in the query, the plannerwill use this information during cardinality estimation. +

Reviewer 2: ```suggestion
a|Allow label inference during cardinality estimation. During cardinality estimation, if the planner can logically deduce that a node has a label that was not explicitly expressed in the query, the planner will use this information during cardinality estimation. +
```

This setting controls to what extent the planner should do that:

* `OFF`: No predicates are inferred.

> * `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes.
> The label corresponding to the smallest number of nodes is used to estimate rows.
> Not inferring more labels yields better results for nodes with several dependent labels.

Reviewer 2: I realise that the [comment](https://github.com/neo-technology/neo4j/pull/25767#discussion_r1626345752) I left on your other PR should have applied to Arne's PR initially.

```suggestion
* `MOST_SELECTIVE_LABEL`: Relationship types are used to infer labels on the relationships' end nodes. The planner will only infer at most one label per node. If more than one label can be inferred for a given node, the planner keeps the most selective one, the one corresponding to the smallest number of nodes in the graph.
```

|Valid values
a|One of [MOST_SELECTIVE_LABEL, OFF].
|Default value

> m|+++OFF+++

Reviewer 2: This is about to change but I'll make sure to come change it
